include/natalie/array_packer/packer.hpp

```diff
@@ -128,9 +128,33 @@ namespace ArrayPacker {
                     });
                     break;
                 }
-                case 'x':
-                    // TODO
+                case 'x': {
+                    // Asterisks have no effect on this directive.
+                    if (token.star)
+                        break;
+
+                    auto null_byte_count = (token.count < 0) ? 1: token.count;
+                    for (int count = 0; count < null_byte_count; count++) {
+                        m_packed.append_char('\0');
+                    }
+                    break;
+                }
+                case 'X': {
+                    // Asterisks have no effect on this directive.
+                    if (token.star)
+                        break;
```

Comment:
Similarly, for this `break`, how doe this make the [spec](https://github.com/seven1m/ruby_spec/blob/6058ec330b3ed47c55c1dd6feef3ff8c33d6a0bb/core/string/unpack/x_spec.rb#L22-L24) pass?

>it "moves the read index back by the remaining bytes when passed the '*' modifier"

What I understand from that spec is this (but maybe I'm wrong):

1. Given the code:
 
   ```
   "abcd".unpack("C3X*C")
   ```

2. `C3` consumes three 8-bit bytes, resulting in:

   ```
   [97, 98, 99]
   ```

   ...and leaving the cursor (m_index) here:

   ```
   "abcd"
       ^
   ```

3. `X*` moves the cursor "back by the remaining bytes". In this case, the remaining bytes is `1`, since we are on the last byte of the string. So the cursor moves back to the `c`, like this:

   ```
   "abcd"
      ^
   ```

4. The last directive of `C` grabs an 8-bit byte, which is the representation of `c`, resulting in:

   ```
   [97, 98, 99, 99]
   ```

I'm curious how the specs are passing given that we are ignoring this directive. :thinking: